provider/pkg/provider/helm_release.go

```diff
@@ -0,0 +1,1193 @@
+	settings.RegistryConfig = registryConfigPath
+	settings.RepositoryConfig = repositoryConfigPath
+	settings.RepositoryCache = repositoryCache
+
+	return &helmReleaseProvider{
+		host:             host,
+		kubeConfig:       kc,
+		helmDriver:       helmDriver,
+		defaultNamespace: namespace,
+		enableSecrets:    enableSecrets,
+		name:             "kubernetes:helmrelease",
+		settings:         settings,
+	}, nil
+}
+
+func debug(format string, a ...interface{}) {
+	logger.V(9).Infof("[DEBUG] %s", fmt.Sprintf(format, a...))
+}
+
+func (r *helmReleaseProvider) getActionConfig(namespace string) (*action.Configuration, error) {
+	conf := new(action.Configuration)
+	if err := conf.Init(r.kubeConfig, namespace, r.helmDriver, debug); err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+func decodeRelease(pm resource.PropertyMap) (*Release, error) {
+	var release Release
+	stripped := pm.MapRepl(nil, mapReplStripSecrets)
+	logger.V(9).Infof("Decoding release: %#v", stripped)
+	if err := mapstructure.Decode(stripped, &release); err != nil {
+		return nil, fmt.Errorf("decoding failure: %w", err)
+	}
+	return &release, nil
+}
+
+func (r *helmReleaseProvider) Check(ctx context.Context, req *pulumirpc.CheckRequest, displayBetaWarning bool) (*pulumirpc.CheckResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("Provider[%s].Check(%s)", r.name, urn)
+
+	if displayBetaWarning {
+		_ = r.host.LogStatus(ctx, diag.Warning, urn, "Helm Release resource is currently in beta and may change. Use in production environments is discouraged.")
+	}
+	// Obtain old resource inputs. This is the old version of the resource(s) supplied by the user as
+	// an update.
+	oldResInputs := req.GetOlds()
+	olds, err := plugin.UnmarshalProperties(oldResInputs, plugin.MarshalOptions{
+		Label: fmt.Sprintf("%s.olds", label), KeepUnknowns: true, SkipNulls: true, KeepSecrets: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Obtain new resource inputs. This is the new version of the resource(s) supplied by the user as
+	// an update.
+	newResInputs := req.GetNews()
+	news, err := plugin.UnmarshalProperties(newResInputs, plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.news", label),
+		KeepUnknowns: true,
+		SkipNulls:    true,
+		KeepSecrets:  true,
+	})
+	if err != nil {
+		return nil, pkgerrors.Wrapf(err, "check failed because malformed resource inputs: %+v", err)
+	}
+
+	logger.V(9).Infof("Decoding new release.")
+	new, err := decodeRelease(news)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.V(9).Infof("Decoding old release.")
+	old, err := decodeRelease(olds)
+	if err != nil {
+		return nil, err
+	}
+
+	if new.ReleaseSpec.Namespace == "" {
+		new.ReleaseSpec.Namespace = r.defaultNamespace
+	}
+
+	if !new.ReleaseSpec.SkipAwait && new.ReleaseSpec.Timeout == 0 {
+		new.ReleaseSpec.Timeout = 300
+	}
+
+	if len(olds.Mappable()) > 0 {
+		adoptOldNameIfUnnamed(new, old)
+		if err = r.helmUpdate(ctx, urn, news, new, old, true); err != nil {
```

Comment:
We are doing a full dry-run update or create in check so we can reliably populate resourceNames/manifest etc. in time for the preview. Doing it in other handlers didn't render the injected input in diffs.